Changes to PlayerData, NetItem, TSPlayer and hooks

AgaSpace opened the pull request:
<!-- Warning: If you create a pull request and wish to remain anonymous, you are highly advised to use Tails (https://tails.boum.org/) or a fresh git environment. We will *not* be able to help with anonymization after your pull request has been created. -->

<!-- Warning: If you do not update the changelog, your pull request will be ignored and eventually closed, without comment, without any support, and without any opinion or interaction from the TShock team. This is your only warning. -->

AgaSpace committed:
Added a new account hook

AgaSpace committed:
Update CHANGELOG.md

AgaSpace committed:
Updated the PlayerData constructor

I added two constructors, one of which saves the old constructor, and the new one has an argument about whether to add items from the config.

I also added a slot check if it is less than 0 for StoreSlot.

AgaSpace committed:
Update TSPlayer.cs

Added the ability to select the cause of death in KillPlayer and DamagePlayer.

Added an xml to send a message in SetPvP

Since GodMode can now be set in Journey mode, we can remove the unnecessary field and add a property whereby we can change the value of GodMode for the player.

Added a property to get the RemoteClient of the player.

Added the ability to get a player's position through a property.

Added an option to update section visibility.

AgaSpace committed:
Removed the unnecessary update of the player's godmode.

AgaSpace committed:
Added a check on the range of possible team set for a player

AgaSpace committed:
Update TSPlayer.cs

AgaSpace committed:
Update NetItem.cs

Added a constructor with an item from Terraria

Added a property that creates a copy of the item, but for Terraria

AgaSpace committed:
Updated classes depending on the update of the PlayerData constructor

AgaSpace committed:
Merge branch 'general-devel' of https://github.com/AgaSpace/TShock into general-devel

AgaSpace committed:
Updating the hooks

* Renamed the hook from "PlayerCommand" to "PrePlayerCommand" and "PostPlayerCommand".
+ Added "AccountGroupUpdate" hook, which is called when trying to use /user group.

AgaSpace committed:
Update Commands.cs

I added i18n when trying to change the player group in the hook "OnAccountGroupChange".

AgaSpace committed:
Logged all my changes

AgaSpace committed:
Update PlayerData.cs

Added the "PlayerData.StoreSlot" method with arguments "int slot" and "NetItem item".

AgaSpace committed:
Update NetItem.cs

Added optional parameters to the NetItem constructor.

AgaSpace committed:
Update Commands.cs

Final edits.
* Fixed the bug with the /user group
* Fixed bug with calling AccountHooks.OnAccountGroupChange hook and returned Handled value.

sgkoishi commented:
*Plugins that use `TSPlayer.DamagePlayer` or `KillPlayer` requires recompile (source level compatible).*
*Plugins that use `TShockAPI.Hooks.PlayerHooks.OnPlayerCommand` requires change.*

punchready reviewed (changes requested):
- The `OnAccountGroupChange` hook should be invoked in `UserManager.SetUserGroup` instead, right now it only fires when a group gets changed via command
- The `PlayerCommand` hook should probably be kept and called like it was before for binary backwards compatibility
- The `UpdateSection` method's `visibility` parameter should be called something like `isLoaded` instead to not be misleading, and the documentation adjusted
- The `DamagePlayer` and `KillPlayer` reason parameters should also be put into an overload instead for binary backwards compatibility

AgaSpace committed:
Updated the "AccountGroupUpdate" hook.

AgaSpace committed:
Split the "PrePlayerCommand" hook into "OnPlayerCommand" and "PrePlayerCommand

AgaSpace committed:
Update TSPlayer.cs

Added overloads for "KillPlayer" and "DamagePlayer"

AgaSpace committed:
Renamed the "visibility" parameter to "isLoaded" in the "TSPlayer.UpdateSection" method

AgaSpace committed:
Update changelog.md

AgaSpace committed:
Update changelog.md

AgaSpace commented:
> * The `OnAccountGroupChange` hook should be invoked in `UserManager.SetUserGroup` instead, right now it only fires when a group gets changed via command
> * The `PlayerCommand` hook should probably be kept and called like it was before for binary backwards compatibility
> * The `UpdateSection` method's `visibility` parameter should be called something like `isLoaded` instead to not be misleading, and the documentation adjusted
> * The `DamagePlayer` and `KillPlayer` reason parameters should also be put into an overload instead for binary backwards compatibility

I wrote it down as you instructed

punchready commented on TShockAPI/TSPlayer.cs:
```diff
@@ -1815,15 +1872,35 @@ public void SendFileTextAsMessage(string file)
 		{
-			NetMessage.SendPlayerHurt(Index, PlayerDeathReason.LegacyDefault(), damage, (new Random()).Next(-1, 1), false, false, 0, -1, -1);
+			NetMessage.SendPlayerHurt(Index, PlayerDeathReason.LegacyDefault(),
+				damage, (new Random()).Next(-1, 1), false, false, 0, -1, -1);
```
This should call the second overload with the default reason.

punchready commented on TShockAPI/TSPlayer.cs:
```diff
@@ -1815,15 +1872,35 @@ public void SendFileTextAsMessage(string file)
 		/// <param name="damage">The amount of damage the player will take.</param>
 		public virtual void DamagePlayer(int damage)
 		{
-			NetMessage.SendPlayerHurt(Index, PlayerDeathReason.LegacyDefault(), damage, (new Random()).Next(-1, 1), false, false, 0, -1, -1);
+			NetMessage.SendPlayerHurt(Index, PlayerDeathReason.LegacyDefault(),
+				damage, (new Random()).Next(-1, 1), false, false, 0, -1, -1);
+		}
+		/// <summary>
+		/// Wounds the player with the given damage.
+		/// </summary>
+		/// <param name="damage">The amount of damage the player will take.</param>
+		/// <param name="reason">Player death reason</param>
+		public virtual void DamagePlayer(int damage, PlayerDeathReason reason)
+		{
+			NetMessage.SendPlayerHurt(Index, reason,
+				damage, (new Random()).Next(-1, 1), false, false, 0, -1, -1);
 		}
 
 		/// <summary>
 		/// Kills the player.
 		/// </summary>
 		public virtual void KillPlayer()
 		{
-			NetMessage.SendPlayerDeath(Index, PlayerDeathReason.LegacyDefault(), 99999, (new Random()).Next(-1, 1), false, -1, -1);
+			NetMessage.SendPlayerDeath(Index, PlayerDeathReason.LegacyDefault(),
+				99999, (new Random()).Next(-1, 1), false, -1, -1);
```
Same here; call the second overload.

punchready commented on TShockAPI/TSPlayer.cs:
```diff
@@ -1840,6 +1919,7 @@ public virtual void SetTeam(int team)
 		/// Sets the player's pvp.
 		/// </summary>
 		/// <param name="mode">The state of the pvp mode.</param>
+		/// <param name="withMsg">Is it necessary to send a message about changing pvp mode</param>
```
The grammar is slightly off, "Whether a chat message about the change should be sent." would be better.

punchready commented on TShockAPI/Hooks/PlayerHooks.cs:
```diff
@@ -449,6 +499,50 @@ public static bool OnPlayerCommand(TSPlayer player, string cmdName, string cmdTe
+				CommandArgs = args
+			};
+
+			args = playerCommandEventArgs.CommandArgs;
```
This line is probably supposed to be after the hook invocation.

punchready commented on TShockAPI/DB/UserManager.cs:
```diff
@@ -183,6 +185,38 @@ public void SetUserGroup(UserAccount account, string group)
+			if (_database.Query("UPDATE Users SET UserGroup = @0 WHERE Username = @1;", group, account.Name) == 0)
+				throw new UserAccountNotExistException(account.Name);
+
+			if (AccountHooks.OnAccountGroupUpdate(account, author, ref grp))
```
This is not how an account update hook should be implemented:
- This still doesn't get invoked when a plugin changes a user's group with the other overload
- This gets invoked after the change has already been made in the database, not allowing proper cancellation
- This throws if canceled?? It should probably either not be cancelable or have a clear return value indicating whether the change was canceled, or cancel by setting the group to the previous group of the user silently

punchready commented on TShockAPI/Hooks/AccountHooks.cs:
```diff
@@ -39,6 +40,26 @@ public AccountCreateEventArgs(UserAccount account)
+		public bool Handled = false;
+
+		public TSPlayer Author { get; private set; }
+		/// <summary>
+		/// Contains only the user name. Use <see cref="UserAccountManager.GetUserAccountByName(string)"/> for more information.
+		/// </summary>
+		public UserAccount Account { get; private set; }
```
This event should either only contain a user account name field or ensure that a full account instance has been fetched to not be confusing to consumers.

AgaSpace replied:
> This event should either only contain a user account name field or ensure that a full account instance has been fetched to not be confusing to consumers.

Should I rename the next property to "AccountName" or should I leave it as it was, which is "Account"?

punchready replied:
`string AccountName` or `UserAccount Account`

AgaSpace committed:
Update TSPlayer.cs

Calling overload methods in "KillPlayer" and "DamagePlayer"

AgaSpace committed:
Update TSPlayer.cs

Updated the documentation for the SetPvP method

AgaSpace committed:
Update PlayerHooks.cs

Updated the PrePlayerCommand event call. (Fixed CommandArgs update not working)

AgaSpace committed:
Update UserManager.cs

Fixed the OnAccountGroupUpdate hook call. (Executed after changing a player's group)

AgaSpace committed:
Update AccountHooks.cs

Updated the Account property. Renamed it to AccountName and set the return type to string.

AgaSpace committed:
Update UserManager.cs

Corrected the change of user group

AgaSpace commented:
@punchready, Any other changes needed for the merge?

punchready commented on TShockAPI/DB/UserManager.cs:
```diff
@@ -183,6 +185,38 @@ public void SetUserGroup(UserAccount account, string group)
+				throw new GroupNotExistsException(group);
+
+			if (AccountHooks.OnAccountGroupUpdate(account, author, ref grp))
+				throw new ArgumentException("The hook blocked the change of the user group.");
```
This should throw a custom exception class at least.

punchready commented on TShockAPI/Hooks/AccountHooks.cs:
```diff
@@ -39,6 +40,23 @@ public AccountCreateEventArgs(UserAccount account)
+	{
+		public bool Handled = false;
+
+		public TSPlayer Author { get; private set; }
```
Again, an account group update event is quite pointless if it only fires when changing a group via command and not from all other calls to the `SetUserGroup` method.

AgaSpace replied:
Then I need to specify that Author can be null (that is, I will add documentation to the Author property that explains this) and then add a hook call to the SetUserGroup overload?

punchready replied:
I would make an abstract `AccountGroupUpdateEventArgs` base class and then have a class for plugin-initiated events like `AccountGroupUpdateByPluginEventArgs` and one for player-initiated events like `AccountGroupUpdateByPlayerEventArgs`.

AgaSpace replied:
> I would make an abstract `AccountGroupUpdateEventArgs` base class and then have a class for plugin-initiated events like `AccountGroupUpdateByPluginEventArgs` and one for player-initiated events like `AccountGroupUpdateByPlayerEventArgs`.

Do I need to pass `TerrariaPlugin` to `AccountGroupUpdateByPluginEventArgs`?

punchready replied:
That's probably too much effort for too little use cases.

sgkoishi replied:
Account group update can only be triggered by either `/user` with `tshock.superadmin.user` permission - superadmin, server console, or plugin. Maintainers can access the database and bypass the hook, and so do the plugins.
What's the use case of this hook instead of using MonoMod?

AgaSpace replied:
> Account group update can only be triggered by either `/user` with `tshock.superadmin.user` permission - superadmin, server console, or plugin. Maintainers can access the database and bypass the hook, and so do the plugins. What's the use case of this hook instead of using MonoMod?

Are you saying that it would still be better to keep 1 class `AccountGroupUpdateEventArgs`?

sgkoishi replied:
> Are you saying that it would still be better to keep 1 class `AccountGroupUpdateEventArgs`?

At least it is confusing for me to tell which is the initiator if the plugins are chained together.

<hr>

If a plugin dev wants to be triggered when `SetUserGroup`, they can just go this (without this PR) - the whole group change event is like saving three lines of code
```csharp
using TShockAPI.DB;

new MonoMod.RuntimeDetour.Hook(
    () => TShock.UserAccounts.SetUserGroup(null, null), 
    (Action<UserAccountManager, UserAccount, string> originalSetGroup, UserAccountManager mgr, UserAccount acc, string a) => {
        Console.WriteLine("Before the group change");
        if (happyEnough && goodLuck)
        {
            originalSetGroup(mgr, acc, a);
        }
        Console.WriteLine("After the group change");
    });
```
now becomes
```csharp
using TShockAPI.DB;

Hooks.AccountHooks.AccountDelete += args => {
        Console.WriteLine("Before the group change");
        if (!happyEnough || !goodLuck)
        {
            args.Handled = true;
        }
        Console.WriteLine("Still before the change, no hook for after the group change");
    };
```

AgaSpace replied:
@sgkoishi, In this way, you can do everything at all. Then why do we need other hooks?
I mean, our way would be more convenient.

AgaSpace replied:
@sgkoishi Generally speaking, do you want me to add `MonoMod.RuntimeDetour.Hook` to the SetUserGroup method? Then how do I pass the "author" of that action?

sgkoishi replied:
I would say the current version is quite ok and enough - the MonoMod one can provide post-event but I couldn't see much usage of it, it really doesn't matter that much if something happens right before/after a group change. The only required thing I think is the pre-event that can cancel it.

punchready reviewed (changes requested):
Account group change event needs to be reworked, otherwise LGTM.

AgaSpace committed:
Update UserManager.cs

+ Added an exception `UserGroupUpdateLockedException`, which is thrown if an attempt to change the user group fails.
+ Added a call of the `OnAccountGroupUpdate` hook to the obsolete SetUserGroup method.
* Changed the documentation of the constructor `UserAccountNotExistException`. (Error in word "passed" ("pasesd"))

AgaSpace committed:
Update RestManager.cs

+ Passed TSRestPlayer to SetUserGroup.

AgaSpace committed:
Update Commands.cs

Changed the exception from `ArgumentException` to `UserGroupUpdateLockedException`.

AgaSpace committed:
Added `AccountGroupUpdateByPluginEventArgs` and `AccountGroupUpdateByPlayerEventArgs` instead of `AccountGroupUpdateEventArgs`.

AgaSpace commented:
Since I have been asked to split this PR into several smaller PRs, I am closing this PR.
